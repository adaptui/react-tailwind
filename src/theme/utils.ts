import { defaults, isString, isUndefined, mergeWith, tcm } from "../utils";

import {
  DefaultTheme,
  ExtendableDefaultTheme,
  PartialDefaultTheme,
} from "./types";

export const extendTheme = <T = ExtendableDefaultTheme>(theme: T): T => {
  return theme;
};

// Based on https://github.com/tailwindlabs/tailwindcss/blob/6ce8864f47061035cd79b2d563eea72467dcb3b7/src/util/resolveConfig.js
function collectExtends(items: ExtendableDefaultTheme[]) {
  return items.reduce((merged, { extend }) => {
    return mergeWith(merged, extend, (mergedValue, extendValue) => {
      if (isUndefined(mergedValue)) {
        return [extendValue];
      }

      if (Array.isArray(mergedValue)) {
        return [extendValue, ...mergedValue];
      }

      return [extendValue, mergedValue];
    });
  }, {});
}

// Based on https://github.com/tailwindlabs/tailwindcss/blob/6ce8864f47061035cd79b2d563eea72467dcb3b7/src/util/resolveConfig.js
export function mergeThemes(themes: PartialDefaultTheme[]) {
  return {
    ...themes.reduce<DefaultTheme>(
      (merged, theme) => defaults(merged, theme),
      {} as DefaultTheme,
    ),

    // In order to resolve n config objects, we combine all of their `extend` properties
    // into arrays instead of objects so they aren't overridden.
    extend: collectExtends(themes),
  };
}

// Based on https://github.com/tailwindlabs/tailwindcss/blob/6ce8864f47061035cd79b2d563eea72467dcb3b7/src/util/resolveConfig.js
export function mergeExtensions({
  extend,
  ...theme
}: DefaultTheme & { extend?: PartialDefaultTheme }) {
  return mergeWith(theme, extend, (themeValue, extendValue) => {
    return mergeWith(
      themeValue,
      ...extendValue,
      (merged: any, value: any[]) => {
        if (isString(merged) && isString(value)) {
          return tcm(merged, value);
        }

        return undefined;
      },
    );
  });
}
