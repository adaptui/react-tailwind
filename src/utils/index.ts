export * from "./assertions";
export * from "./createContext";
export * from "./debounce";
export * from "./global-types";
export * from "./object";
export * from "./react";
export * from "./tailwindMerge";
export * from "./types";
