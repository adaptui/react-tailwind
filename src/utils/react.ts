import * as React from "react";

import { isFunction } from "./assertions";
import { cx } from "./tailwindMerge";
import {
  As,
  ComponentWithAs,
  Dict,
  PropsWithAs,
  RenderPropType,
} from "./types";

/**
 * @template Props Component Props
 * @template RefProp HTML intrinsic type, eg: HTMLDivElement
 * @template DefaultType string, eg: "div"
 * @param component
 */
export function forwardRefWithAs<
  Props,
  RefProp = any,
  DefaultType extends As = any,
>(
  component: React.ForwardRefRenderFunction<
    RefProp,
    PropsWithAs<Props, DefaultType>
  >,
) {
  return React.forwardRef(component) as unknown as ComponentWithAs<
    Props,
    DefaultType
  >;
}

// From Chakra Utils
export function runIfFn<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  return isFunction(valueOrFn) ? valueOrFn(...args) : valueOrFn;
}

export function runIfFnChildren<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  // Know bug but this is the only way to make it work
  // {({ isChecked }) => {
  //   return (
  //     <>
  //       <SwitchText data-testid="testid-mode">
  //         {isChecked ? DARK_MODE : LIGHT_MODE}
  //       </SwitchText>
  //     </>
  //   );
  // }}
  // Need a wrapper to get the children when the children is a function
  // @ts-ignore
  return isFunction(valueOrFn) ? valueOrFn(...args).props.children : valueOrFn;
}

/**
 * Gets only the valid children of a component,
 * and ignores any nullish or falsy child.
 *
 * @param children the children
 */
export function getValidChildren(children: React.ReactNode) {
  return React.Children.toArray(children).filter(child =>
    React.isValidElement(child),
  ) as React.ReactElement[];
}

// Merge library & user prop
export const passProps = (icon: RenderPropType, props?: Dict) => {
  return React.isValidElement(icon)
    ? React.cloneElement(icon, {
        ...props,
        ...icon.props,
        className: cx(props?.className, icon.props.className),
      })
    : runIfFn(icon, { ...props });
};

// Add a11y to the icon passed
export const withIconA11y = (icon: RenderPropType, props?: Dict) => {
  return passProps(icon, {
    role: "img",
    focusable: false,
    "aria-hidden": true,
    ...props,
  });
};

export const getComponentProps = <T, P>(
  componentMaps: Dict<string>,
  children: T,
  props: P,
) => {
  const normalizedChildren = runIfFnChildren(children, props);
  const validChildren = getValidChildren(normalizedChildren);
  const componentProps: Dict = {};
  const finalChildren: React.ReactNode[] = [];

  validChildren.forEach(child => {
    // @ts-ignore
    if (componentMaps[child.type.displayName]) {
      componentProps[
        // @ts-ignore
        componentMaps[child.type.displayName]
      ] = child.props;
    } else {
      finalChildren.push(child);
    }
  });

  return { componentProps, finalChildren };
};
