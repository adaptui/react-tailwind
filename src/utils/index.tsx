import * as React from "react";
import { createTailwindMerge } from "tailwind-merge";

import { cx } from "../utils";

import {
  As,
  ComponentWithAs,
  Dict,
  PropsWithAs,
  RenderPropType,
} from "./types";

export interface CreateContextOptions {
  /**
   * If `true`, React will throw if context is `null` or `undefined`
   * In some cases, you might want to support nested context, so you can set it to `false`
   */
  strict?: boolean;
  /**
   * Error message to throw if the context is `undefined`
   */
  errorMessage?: string;
  /**
   * The display name of the context
   */
  name?: string;
}

export type CreateContextReturn<T> = [
  React.Provider<T>,
  () => T,
  React.Context<T>,
];

/**
 * Creates a named context, provider, and hook.
 *
 * @param options create context options
 */
export function createContext<ContextType>(options: CreateContextOptions = {}) {
  const {
    strict = true,
    errorMessage = "useContext: `context` is undefined. Seems you forgot to wrap component within the Provider",
    name,
  } = options;

  const Context = React.createContext<ContextType | undefined>(undefined);

  Context.displayName = name;

  function useContext() {
    const context = React.useContext(Context);

    if (!context && strict) {
      throw new Error(errorMessage);
    }

    return context;
  }

  return [
    Context.Provider,
    useContext,
    Context,
  ] as CreateContextReturn<ContextType>;
}

// Function assertions
export function isFunction(value: any): value is Function {
  return typeof value === "function";
}

// From Chakra Utils
export function runIfFn<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  return isFunction(valueOrFn) ? valueOrFn(...args) : valueOrFn;
}

export function runIfFnChildren<T, U>(
  valueOrFn: T | ((...fnArgs: U[]) => T),
  ...args: U[]
): T {
  // Know bug but this is the only way to make it work
  // {({ isChecked }) => {
  //   return (
  //     <>
  //       <SwitchText data-testid="testid-mode">
  //         {isChecked ? DARK_MODE : LIGHT_MODE}
  //       </SwitchText>
  //     </>
  //   );
  // }}
  // Need a wrapper to get the children when the children is a function
  // @ts-ignore
  return isFunction(valueOrFn) ? valueOrFn(...args).props.children : valueOrFn;
}

/**
 * Gets only the valid children of a component,
 * and ignores any nullish or falsy child.
 *
 * @param children the children
 */
export function getValidChildren(children: React.ReactNode) {
  return React.Children.toArray(children).filter(child =>
    React.isValidElement(child),
  ) as React.ReactElement[];
}

// SSR check
export function canUseDOM() {
  return !!(
    typeof window !== "undefined" &&
    window.document &&
    window.document.createElement
  );
}

export const isBrowser = canUseDOM();

// Array assertions
export function isArray<T>(value: any): value is Array<T> {
  return Array.isArray(value);
}

// Object assertions
export const isObject = (value: any): value is Dict => {
  const type = typeof value;

  return value != null && type === "object" && !isArray(value);
};

export const objectKeys = <T extends Dict>(obj: T) =>
  Object.keys(obj) as unknown as (keyof T)[];

// Corresponds to 10 frames at 60 Hz.
// A few bytes payload overhead when lodash/debounce is ~3 kB and debounce ~300 B.
export const debounce = (func: Function, wait = 166) => {
  let timeout: any;
  function debounced(this: any, ...args: any) {
    // eslint-disable-next-line consistent-this
    const that = this;
    const later = () => {
      func.apply(that, args);
    };
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
  }

  debounced.clear = () => {
    clearTimeout(timeout);
  };

  return debounced;
};

// String assertions
export function isString(value: any): value is string {
  return Object.prototype.toString.call(value) === "[object String]";
}

// Merge library & user prop
export const passProps = (icon: RenderPropType, props?: Dict) => {
  return React.isValidElement(icon)
    ? React.cloneElement(icon, {
        ...props,
        ...icon.props,
        className: cx(props?.className, icon.props.className),
      })
    : runIfFn(icon, { ...props });
};

// Add a11y to the icon passed
export const withIconA11y = (icon: RenderPropType, props?: Dict) => {
  return passProps(icon, {
    role: "img",
    focusable: false,
    "aria-hidden": true,
    ...props,
  });
};

export function isUndefined(value: any): value is undefined {
  return typeof value === "undefined" || value === undefined;
}

export const tcm = createTailwindMerge(getDefaultConfig => {
  const defaultConfig = getDefaultConfig();

  return {
    ...defaultConfig,
    classGroups: {
      ...defaultConfig.classGroups,
      "font-size": [
        ...defaultConfig.classGroups["font-size"],
        // Defined custom text presets
        {
          text: ["cxs", "paragraph-cxs", "paragraph-sm", "2base"],
        },
      ],
    },
  };
});

export const getComponentProps = <T, P>(
  componentMaps: Dict<string>,
  children: T,
  props: P,
) => {
  const normalizedChildren = runIfFnChildren(children, props);
  const validChildren = getValidChildren(normalizedChildren);
  const componentProps: Dict = {};
  const finalChildren: React.ReactNode[] = [];

  validChildren.forEach(child => {
    // @ts-ignore
    if (componentMaps[child.type.displayName]) {
      componentProps[
        // @ts-ignore
        componentMaps[child.type.displayName]
      ] = child.props;
    } else {
      finalChildren.push(child);
    }
  });

  return { componentProps, finalChildren };
};

/**
 * @template Props Component Props
 * @template RefProp HTML intrinsic type, eg: HTMLDivElement
 * @template DefaultType string, eg: "div"
 * @param component
 */
export function forwardRefWithAs<
  Props,
  RefProp = any,
  DefaultType extends As = any,
>(
  component: React.ForwardRefRenderFunction<
    RefProp,
    PropsWithAs<Props, DefaultType>
  >,
) {
  return React.forwardRef(component) as unknown as ComponentWithAs<
    Props,
    DefaultType
  >;
}

export * from "./types";
export { cx } from "@renderlesskit/react";
