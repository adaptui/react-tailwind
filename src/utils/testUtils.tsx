import {
  render as RtlRender,
  RenderOptions,
  waitFor,
  RenderResult,
} from "@testing-library/react";
import * as React from "react";
import { RunOptions } from "axe-core";
import { configureAxe } from "jest-axe";

// @ts-ignore
import tailwindConfig from "../../tailwind.config";
import { RenderlesskitProvider } from "../theme";
export * from "@testing-library/react";

type Render = (
  children: React.ReactNode,
  options?: RenderOptions,
) => RenderResult;

export const render: Render = (children, options = {}) => {
  return RtlRender(
    <RenderlesskitProvider tailwindConfig={tailwindConfig}>
      {children}
    </RenderlesskitProvider>,
    options,
  );
};

export const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

type UI = Parameters<typeof render>[0] | Element;
type TestA11YOptions = RenderOptions & {
  axeOptions?: RunOptions;
};

export const testA11y = async (
  ui: UI,
  { axeOptions, ...options }: TestA11YOptions = {},
) => {
  jest.useRealTimers();

  const container = React.isValidElement(ui)
    ? render(ui, options).container
    : ui;

  // @ts-ignore
  const results = await axe(container, axeOptions);

  expect(results).toHaveNoViolations();
};

const DELAY = 0;
const LOAD_IMAGE = "load.png";
const ERROR_IMAGE = "error.png";
const orignalImage = window.Image;

export const mockImage = (loadState: string) => {
  jest.useFakeTimers();

  (window.Image as unknown) = class MockImage {
    onload: () => void = () => {};
    onerror: () => void = () => {};
    src: string = "";
    constructor() {
      if (loadState === LOAD_IMAGE) {
        setTimeout(() => {
          this.onload();
        }, DELAY);
      }
      if (loadState === ERROR_IMAGE) {
        setTimeout(() => {
          this.onerror();
        }, DELAY);
      }
      return this;
    }
  };
};

mockImage.DELAY = DELAY;
mockImage.LOAD_IMAGE = LOAD_IMAGE;
mockImage.ERROR_IMAGE = ERROR_IMAGE;
mockImage.ERROR_IMAGE = ERROR_IMAGE;
mockImage.load = () => mockImage(LOAD_IMAGE);
mockImage.error = () => mockImage(ERROR_IMAGE);
mockImage.restoreMock = () => (window.Image = orignalImage);
mockImage.advanceTimer = async () => {
  return await waitFor(() => {
    jest.advanceTimersByTime(mockImage.DELAY);
  });
};
