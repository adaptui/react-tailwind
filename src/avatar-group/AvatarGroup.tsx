import { Avatar } from "../avatar";
import { forwardRefWithAs } from "../utils";

import { AvatarGroupContextProvider } from "./__utils";
import { AvatarGroupProps, useAvatarGroupProps } from "./AvatarGroupProps";
import { AvatarGroupWrapper } from "./AvatarGroupWrapper";

export const AvatarGroup = forwardRefWithAs<
  AvatarGroupProps,
  HTMLDivElement,
  "div"
>((props, ref) => {
  const { childrenWithinMax, excessChildrenCount, uiProps, wrapperProps } =
    useAvatarGroupProps(props);

  return (
    <AvatarGroupWrapper ref={ref} {...uiProps} {...wrapperProps}>
      <AvatarGroupContextProvider {...uiProps}>
        {childrenWithinMax}
        {excessChildrenCount > 0 ? (
          <Avatar
            name={excessChildrenCount.toString()}
            data-testid="testid-excess_label"
          />
        ) : null}
      </AvatarGroupContextProvider>
    </AvatarGroupWrapper>
  );
});

AvatarGroup.displayName = "AvatarGroup";
