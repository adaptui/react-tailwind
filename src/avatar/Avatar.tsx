import * as React from "react";
import { twMerge as cx } from "tailwind-merge";

import { useTheme } from "../theme";
import { Box, BoxProps } from "../box";
import { GenericAvatar } from "../icons";
import { AvatarName } from "./AvatarName";
import { AvatarIcon } from "./AvatarIcon";
import { AvatarImage } from "./AvatarImage";
import { AvatarBadge } from "./AvatarBadge";
import { useImage } from "../hooks/useImage";
import { useAvatarGroup } from "./AvatarGroup";
import { createContext, runIfFn } from "../utils";
import { forwardRefWithAs, RenderProp } from "../utils/types";

export type AvatarInitialProps = {
  /**
   * How large should avatar be?
   * @default "md"
   */
  size?: keyof Renderlesskit.GetThemeValue<"avatar", "size">;
  /**
   * URL for the avatar image
   */
  src?: string;
  /**
   * Name prop used for `alt` & calculate placeholder initials.
   */
  name?: string;
  /**
   * Defines loading strategy
   */
  loading?: "eager" | "lazy";
  /**
   * Function to get the initials to display
   */
  getInitials?: (name?: string) => string | undefined;
  /**
   * Function called when image failed to load
   */
  onError?: () => void;
  /**
   * The default avatar used as fallback when `name`, and `src`
   * is not specified.
   */
  fallback?: React.ReactNode;
  /**
   * If `true`, the `Avatar` will show a `border` around it.
   *
   * Best for a group of avatars
   */
  showBorder?: boolean;
  /**
   * Color of the `border` to match it's parent background.
   */
  borderColor?: string;
  /**
   * Shows AvatarBadge with the given type
   */
  status?: keyof Renderlesskit.GetThemeValue<"avatar", "status">;
  /**
   * Position for the AvatarBadge
   * @default "bottom-right"
   */
  position?: keyof Renderlesskit.GetThemeValue<"avatar", "badge", "position">;
};

export type AvatarContext = AvatarInitialProps & { showFallback: boolean };

const [AvatarProvider, useAvatarContext] = createContext<AvatarContext>({
  name: "AvatarContext",
  strict: false,
});

export { useAvatarContext };

type AvatarRenderProps = RenderProp<AvatarContext>;

export type AvatarProps = Omit<BoxProps, "onError"> &
  AvatarInitialProps &
  AvatarRenderProps;

export const Avatar = forwardRefWithAs<AvatarProps, HTMLDivElement, "div">(
  (props, ref) => {
    const {
      name,
      src,
      size,
      onError,
      fallback,
      getInitials = initials,
      loading,
      status,
      position,
      showBorder,
      borderColor,
      children,
      className,
      ...rest
    } = props;

    const group = useAvatarGroup();
    const _size = size || group?.size || "md";
    const _showBorder = showBorder || group?.showBorder || false;
    const _borderColor = borderColor || group?.borderColor || false;
    const theme = useTheme();
    const avatarStyles = cx(
      theme.avatar.base,
      theme.avatar.size[_size],
      _showBorder ? theme.avatar.border.width[_size] : "",
      _showBorder && _borderColor ? _borderColor : theme.avatar.border.color,
      className,
    );

    /**
     * Use the image hook to only show the image when it has loaded
     */
    const imageStatus = useImage({ src, onError });
    const hasLoaded = imageStatus === "loaded";

    /**
     * Fallback avatar applies under 2 conditions:
     * - If `src` was passed and the image has not loaded or failed to load
     * - If `src` wasn't passed
     *
     * In this case, we'll show either the name avatar or default avatar
     */
    const showFallback = !src || !hasLoaded;

    const context: AvatarContext = React.useMemo(
      () => ({
        size: _size,
        src,
        name,
        loading,
        getInitials,
        onError,
        fallback,
        status,
        position,
        showFallback,
      }),
      [
        _size,
        src,
        name,
        loading,
        getInitials,
        onError,
        fallback,
        status,
        position,
        showFallback,
      ],
    );

    return (
      <AvatarProvider value={context}>
        <Box
          ref={ref}
          className={avatarStyles}
          {...rest}
          data-testid="testid-avatar_children"
        >
          {children ? (
            runIfFn(children, { ...context, getInitials })
          ) : (
            <>
              <AvatarContents />
              {status ? <AvatarBadge /> : null}
            </>
          )}
        </Box>
      </AvatarProvider>
    );
  },
);

Avatar.displayName = "Avatar";

export function initials(name?: string) {
  if (!name) return;
  const [firstName, lastName] = name.split(" ");
  return firstName && lastName
    ? `${firstName.charAt(0)}${lastName.charAt(0)}`
    : firstName.charAt(0);
}

export const AvatarContents = () => {
  const { showFallback, name, fallback } = useAvatarContext();
  return (
    <>
      {!showFallback ? (
        <AvatarImage />
      ) : name ? (
        <AvatarName />
      ) : (
        <AvatarIcon>{fallback ? fallback : <GenericAvatar />}</AvatarIcon>
      )}
    </>
  );
};
