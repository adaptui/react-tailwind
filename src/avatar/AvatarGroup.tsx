import * as React from "react";
import { cx } from "@renderlesskit/react";

import { useTheme } from "../theme";
import { Box, BoxProps } from "../box";
import { forwardRefWithAs } from "../utils/types";
import { createContext, getValidChildren } from "../utils";
import { Avatar, AvatarContents, AvatarProps } from "./Avatar";

export type AvatarGroupContext = Pick<
  AvatarProps,
  "size" | "showBorder" | "borderColor"
> & {};

const [AvatarGroupProvider, useAvatarGroup] = createContext<AvatarGroupContext>(
  {
    strict: false,
    name: "AvatarGroupProvider",
  },
);

export { useAvatarGroup };

export type AvatarGroupProps = BoxProps &
  AvatarGroupContext & { limit?: number };

export const AvatarGroup = forwardRefWithAs<
  AvatarGroupProps,
  HTMLDivElement,
  "div"
>((props, ref) => {
  const {
    limit = 100,
    size = "md",
    showBorder = true,
    borderColor,
    className,
    children,
    ...rest
  } = props;
  const theme = useTheme();
  const context = React.useMemo(() => ({ size, showBorder, borderColor }), [
    size,
    showBorder,
    borderColor,
  ]);

  const validChildren = getValidChildren(children);

  /**
   * Get the avatars within the max
   */
  const childrenWithinMax = limit
    ? validChildren.slice(0, limit)
    : validChildren;

  /**
   * Get the remaining avatar count
   */
  const excess = limit != null && validChildren.length - limit;

  return (
    <AvatarGroupProvider value={context}>
      <Box
        ref={ref}
        role="group"
        aria-label="Avatar group"
        className={cx(theme.avatar.group.base, className)}
        {...rest}
      >
        {childrenWithinMax}
        <AvatarExcess
          size={size}
          excess={excess}
          {...validChildren[limit].props}
        />
      </Box>
    </AvatarGroupProvider>
  );
});

const AvatarExcess = ({
  excess,
  size = "md",
  ...props
}: AvatarProps & {
  excess: number | false;
}) => {
  const theme = useTheme();

  return excess > 0 ? (
    <Avatar {...props}>
      <>
        <AvatarContents />
        <Box className={theme.avatar.group.excess.bg} />
        <Box
          data-testid="testid-excess_label"
          className={cx(
            theme.avatar.group.excess.text.base,
            theme.avatar.group.excess.text.size[size],
          )}
        >
          +{size === "xs" ? "" : excess}
        </Box>
      </>
    </Avatar>
  ) : null;
};
