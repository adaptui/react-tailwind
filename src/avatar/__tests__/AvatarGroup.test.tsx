import { screen } from "@testing-library/react";

import { render } from "../../utils/testUtils";
import { Avatar } from "../Avatar";
import { AvatarGroup } from "../AvatarGroup";

describe("<AvatarGroup />", () => {
  expect.assertions(1);

  it("should render correctly", () => {
    const { asFragment } = render(
      <AvatarGroup>
        <Avatar />
        <Avatar />
      </AvatarGroup>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should pass the size to all avatars", () => {
    const { asFragment } = render(
      <AvatarGroup size="lg">
        <Avatar />
        <Avatar />
      </AvatarGroup>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should be able to overwrite specific avatar size", () => {
    const { asFragment } = render(
      <AvatarGroup size="lg">
        <Avatar />
        <Avatar size="xs" />
      </AvatarGroup>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should not render truncated avatar counter if avatar count is less than max", () => {
    render(
      <AvatarGroup max={5}>
        <Avatar />
        <Avatar />
        <Avatar />
      </AvatarGroup>,
    );

    const truncateCircle = screen.queryByTestId("testid-excess_label");
    expect(truncateCircle).not.toBeInTheDocument();

    const group = screen.getByRole("group", {
      name: /avatar group/i,
    });

    expect(group.childElementCount).toEqual(3);
  });

  it("should render truncated avatar counter if avatar count is greater than max", () => {
    render(
      <AvatarGroup max={2}>
        <Avatar />
        <Avatar />
        <Avatar />
      </AvatarGroup>,
    );

    const truncateCircle = screen.queryByTestId("testid-excess_label");
    expect(truncateCircle).toHaveTextContent("1");

    const group = screen.getByRole("group", {
      name: /avatar group/i,
    });

    expect(group.childElementCount).toEqual(3);
  });

  it("should not render truncated avatars counter if max is equal to avatars count", () => {
    render(
      <AvatarGroup max={2}>
        <Avatar />
        <Avatar />
      </AvatarGroup>,
    );

    const truncateCircle = screen.queryByTestId("testid-excess_label");
    expect(truncateCircle).not.toBeInTheDocument();

    const group = screen.getByRole("group", {
      name: /avatar group/i,
    });

    expect(group.childElementCount).toEqual(2);
  });
});
