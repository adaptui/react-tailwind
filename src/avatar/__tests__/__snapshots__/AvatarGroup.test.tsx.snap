// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvatarGroup /> should be able to overwrite specific avatar size 1`] = `
<DocumentFragment>
  <div
    aria-label="Avatar group"
    class="inline-flex items-center -space-x-1.5"
    role="group"
  >
    <div
      class="relative inline-flex items-center justify-center rounded-full flex-shrink-0 bg-gray-100 h-8 w-8 ring-2 ring-white"
      data-testid="testid-avatar_children"
    >
      <span
        aria-label="Avatar Icon"
        class="inline-flex text-gray-500 text-xs"
      >
        <svg
          aria-hidden="true"
          class="w-[1em] h-[1em] inline-block leading-[1em] flex-shrink-0 text-current align-middle"
          data-testid="testid-icon"
          focusable="false"
          role="img"
          viewBox="0 0 12 12"
        >
          <path
            d="M6 5c1.1046 0 2-.8954 2-2s-.8954-2-2-2-2 .8954-2 2 .8954 2 2 2zM6 7c-.6566 0-1.3068.1035-1.9134.3045s-1.1578.4956-1.6221.867C1.5268 8.9218 1 9.9392 1 11h10c0-1.0609-.5268-2.0783-1.4645-2.8284C8.5978 7.4214 7.3261 7 6 7z"
            fill="transparent"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.3"
          />
        </svg>
      </span>
    </div>
    <div
      class="relative inline-flex items-center justify-center rounded-full flex-shrink-0 bg-gray-100 h-4 w-4 ring-1 ring-white"
      data-testid="testid-avatar_children"
    >
      <span
        aria-label="Avatar Icon"
        class="inline-flex text-gray-500 text-[0.625rem]"
      >
        <svg
          aria-hidden="true"
          class="w-[1em] h-[1em] inline-block leading-[1em] flex-shrink-0 text-current align-middle"
          data-testid="testid-icon"
          focusable="false"
          role="img"
          viewBox="0 0 12 12"
        >
          <path
            d="M6 5c1.1046 0 2-.8954 2-2s-.8954-2-2-2-2 .8954-2 2 .8954 2 2 2zM6 7c-.6566 0-1.3068.1035-1.9134.3045s-1.1578.4956-1.6221.867C1.5268 8.9218 1 9.9392 1 11h10c0-1.0609-.5268-2.0783-1.4645-2.8284C8.5978 7.4214 7.3261 7 6 7z"
            fill="transparent"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.3"
          />
        </svg>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AvatarGroup /> should pass the size to all avatars 1`] = `
<DocumentFragment>
  <div
    aria-label="Avatar group"
    class="inline-flex items-center -space-x-1.5"
    role="group"
  >
    <div
      class="relative inline-flex items-center justify-center rounded-full flex-shrink-0 bg-gray-100 h-8 w-8 ring-2 ring-white"
      data-testid="testid-avatar_children"
    >
      <span
        aria-label="Avatar Icon"
        class="inline-flex text-gray-500 text-xs"
      >
        <svg
          aria-hidden="true"
          class="w-[1em] h-[1em] inline-block leading-[1em] flex-shrink-0 text-current align-middle"
          data-testid="testid-icon"
          focusable="false"
          role="img"
          viewBox="0 0 12 12"
        >
          <path
            d="M6 5c1.1046 0 2-.8954 2-2s-.8954-2-2-2-2 .8954-2 2 .8954 2 2 2zM6 7c-.6566 0-1.3068.1035-1.9134.3045s-1.1578.4956-1.6221.867C1.5268 8.9218 1 9.9392 1 11h10c0-1.0609-.5268-2.0783-1.4645-2.8284C8.5978 7.4214 7.3261 7 6 7z"
            fill="transparent"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.3"
          />
        </svg>
      </span>
    </div>
    <div
      class="relative inline-flex items-center justify-center rounded-full flex-shrink-0 bg-gray-100 h-8 w-8 ring-2 ring-white"
      data-testid="testid-avatar_children"
    >
      <span
        aria-label="Avatar Icon"
        class="inline-flex text-gray-500 text-xs"
      >
        <svg
          aria-hidden="true"
          class="w-[1em] h-[1em] inline-block leading-[1em] flex-shrink-0 text-current align-middle"
          data-testid="testid-icon"
          focusable="false"
          role="img"
          viewBox="0 0 12 12"
        >
          <path
            d="M6 5c1.1046 0 2-.8954 2-2s-.8954-2-2-2-2 .8954-2 2 .8954 2 2 2zM6 7c-.6566 0-1.3068.1035-1.9134.3045s-1.1578.4956-1.6221.867C1.5268 8.9218 1 9.9392 1 11h10c0-1.0609-.5268-2.0783-1.4645-2.8284C8.5978 7.4214 7.3261 7 6 7z"
            fill="transparent"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.3"
          />
        </svg>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AvatarGroup /> should render correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="Avatar group"
    class="inline-flex items-center -space-x-1.5"
    role="group"
  >
    <div
      class="relative inline-flex items-center justify-center rounded-full flex-shrink-0 bg-gray-100 h-6 w-6 ring-2 ring-white"
      data-testid="testid-avatar_children"
    >
      <span
        aria-label="Avatar Icon"
        class="inline-flex text-gray-500 text-xs"
      >
        <svg
          aria-hidden="true"
          class="w-[1em] h-[1em] inline-block leading-[1em] flex-shrink-0 text-current align-middle"
          data-testid="testid-icon"
          focusable="false"
          role="img"
          viewBox="0 0 12 12"
        >
          <path
            d="M6 5c1.1046 0 2-.8954 2-2s-.8954-2-2-2-2 .8954-2 2 .8954 2 2 2zM6 7c-.6566 0-1.3068.1035-1.9134.3045s-1.1578.4956-1.6221.867C1.5268 8.9218 1 9.9392 1 11h10c0-1.0609-.5268-2.0783-1.4645-2.8284C8.5978 7.4214 7.3261 7 6 7z"
            fill="transparent"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.3"
          />
        </svg>
      </span>
    </div>
    <div
      class="relative inline-flex items-center justify-center rounded-full flex-shrink-0 bg-gray-100 h-6 w-6 ring-2 ring-white"
      data-testid="testid-avatar_children"
    >
      <span
        aria-label="Avatar Icon"
        class="inline-flex text-gray-500 text-xs"
      >
        <svg
          aria-hidden="true"
          class="w-[1em] h-[1em] inline-block leading-[1em] flex-shrink-0 text-current align-middle"
          data-testid="testid-icon"
          focusable="false"
          role="img"
          viewBox="0 0 12 12"
        >
          <path
            d="M6 5c1.1046 0 2-.8954 2-2s-.8954-2-2-2-2 .8954-2 2 .8954 2 2 2zM6 7c-.6566 0-1.3068.1035-1.9134.3045s-1.1578.4956-1.6221.867C1.5268 8.9218 1 9.9392 1 11h10c0-1.0609-.5268-2.0783-1.4645-2.8284C8.5978 7.4214 7.3261 7 6 7z"
            fill="transparent"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.3"
          />
        </svg>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
